062-topic-replication-factor-change.md

```diff
@@ -0,0 +1,254 @@
+In case of a single-rack deployment, rack goals are not configured and the Topic Operator will automatically enable `skip_rack_awareness_check`.
+
+There is also a `replication_throttle` parameter, which is an upper bound on the bandwidth used to move replicas (bytes per second).
+This is already configurable in `Kafka.spec.cruiseControl.config`, and it also applies to all task executions, including replication factor changes.
+
+Cruise Control defaults are used for all the other parameters.
+Some of them can be configured from `Kafka` resource (i.e. default goals).
+
+#### User tasks
+
+The `ReplicasChangeManager` calls the `user_tasks` endpoint to get the task status, passing the `User-Task-ID` header value returned by the `topic_configuration` response.
+
+Example:
+
+```sh
+$ curl -s "localhost:9090/kafkacruisecontrol/user_tasks?user_task_ids=5344ca89-351f-4565-8d0f-9aade00e053d&json=true" | jq
+{
+  "userTasks": [
+    {
+      "Status": "Completed",
+      "ClientIdentity": "127.0.0.1",
+      "RequestURL": "POST /kafkacruisecontrol/topic_configuration?topic=my-topic&replication_factor=2&skip_rack_awareness_check=true&dryrun=false&json=true",
+      "UserTaskId": "5344ca89-351f-4565-8d0f-9aade00e053d",
+      "StartMs": "1701961135978"
+    }
+  ],
+  "version": 1
+}
+```
+
+### Corner cases
+
+Here we analyze some events that may happen during the replication factor change execution, and how they are handled by the Topic Operator.
+
+#### Topic deleted
+
+If it's deleted in Kubernetes, the Topic Operator also deletes the topic in Kafka, and logs an error for the ongoing replication factor change, unless the `KafkaTopic` is paused or unmanaged.
+If it's deleted in Kafka, the Topic Operator recreates the topic in Kafka with the target replication factor, unless the `KafkaTopic` is paused or unmanaged.
+
+#### Cruise Control crashes
+
+Pending replication factor change requests are not durable, which means they are lost if Cruise Control crashes or is restarted after accepting the request.
+In that case, the Topic Operator gets an exception while polling, removes the pending change from its in-memory cache, and retries until Cruise Control comes back.
+
+#### Multiple changes for the same topic
+
+A reconciliation batch can only contain one event related to a specific topic, other events for the same topic are pushed back to the from of the queue.
+It is still possible that one or more replication factor changes for the same topic are requested while another one is already ongoing.
+In this case, the Topic Operator ignores intermediate changes, and only reconciles the last update when the ongoing change in completed.
+
+#### Topic Operator crashes
+
+In this case, there may be some pending and some ongoing replication factor changes.
+All pending and ongoing changes are recovered by the periodic reconciliation as soon as the Topic Operator restarts.
+
+## Affected/not affected projects
```

Comment:
Are we going to factor out the Cruise Control API components and reuse them in the TO?
We already have logic for issuing API REST requests, i.e. running tasks and checking their status.
I think it would be really beneficial to the Strimzi codebase and have an impact on the Kafka rebalance related stuff as well. I would leverage this proposal to do such a refactoring.

Reply:
I agree, but I would prefer to do this refactoring as a separate PR, as it is not so straightforward as it may seem, and it would also complicate the implementation review.

Reply:
> doing this to simplify the code and having a common and shared CC REST API implementation without any duplication

I agree on that. I just don't want to do it in the scope of this proposal to avoid over complicating the implementation's review. The HTTP integration code is minimal for this feature, which makes the review significantly easier. Once this is implemented with all the tests, we can open a dedicated PR to create a shared implementation. Wdyt?

Reply:
This sounds something that should be discussed within this proposal. (Just to be clear, one proposal != one PR. The proposal should include how things will look like, but might not necessarily be implemented in a single PR.)

Reply:
also ...

> The HTTP integration code is minimal for this feature, which makes the review significantly easier

I could totally understand this if the addition was made on top of the already existing CC REST API component we already have. Without that, you are going to write/duplicate the underneath HTTP logic which would force reviewers to check more code instead.

Reply:
ok, will rephrase the, and we will need two or more PRs

Reply:
I agree with Paolo that 2 PRs do not sound easier to review than one PR here. But I do not think this matters for the proposal. That should describe the final state and not how any PRs will be done.